src/btree/bt_read.c

```diff
@@ -255,89 +249,65 @@ __instantiate_lookaside(WT_SESSION_IMPL *session, WT_REF *ref)
-        WT_ERR(
-          las_cursor->get_key(las_cursor, &las_btree_id, &las_key, &las_timestamp, &las_txnid));
-
-        /* Stop before crossing over to the next btree. */
-        if (las_btree_id != S2BT(session)->id)
-            break;
-
-        WT_ERR(__wt_compare(session, NULL, &las_key, &page_las->min_las_key, &cmp));
-        /* We got a key that is less than the minimum key of the page. */
-        if (cmp < 0)
-            break;
+    for (i = 0; i < page_las->birthmarks_cnt; ++i) {
+        if (page_las->birthmarks[i].txnid != WT_TXN_ABORTED) {
+            WT_ERR(__create_birthmark_upd(session, &page_las->birthmarks[i], &size, &upd));
+            WT_ERR(__wt_buf_set(session, &las_key, page_las->birthmarks[i].key.data,
+              page_las->birthmarks[i].key.size));
+            birthmark_record = true;
+        } else {
+            las_cursor->set_key(
+              las_cursor, las_btree_id, &page_las->birthmarks[i].key, WT_TS_MAX, WT_TXN_MAX);
+            WT_ERR(las_cursor->search_near(las_cursor, &exact));
 
-        WT_ERR(__wt_compare(session, NULL, &las_key, &page_las->max_las_key, &cmp));
-        /* We reached the end of the key of the LAS records related to the current page. */
-        if (cmp > 0)
-            break;
+            /*
+             * Check whether the search_near returned the same key. If it returns the next key,
+             * continue with the previous record (assuming there is a previous record).
+             */
+            if (exact > 0 && (ret = las_cursor->prev(las_cursor)) != 0)
+                break;
+            WT_ERR(
+              las_cursor->get_key(las_cursor, &las_btree_id, &las_key, &las_timestamp, &las_txnid));
 
-        instantiated_cnt++;
-        if (page_las->birthmarks_cnt > 0) {
-            for (birthmarkp = page_las->birthmarks + birthmark_cnt;
-                 birthmark_cnt < page_las->birthmarks_cnt; birthmark_cnt++, birthmarkp++) {
-                if (birthmarkp->instantiated)
-                    continue;
-
-                WT_ERR(__wt_compare(session, NULL, &las_key, &birthmarkp->key, &cmp));
-                if (cmp == 0) {
-                    WT_ERR(__create_birthmark_upd(session, birthmarkp, &size, &upd));
-                    birthmark_record = true;
-                    break;
-                } else if (cmp < 0)
-                    /* LAS key is less than birthmark key, so no birthmark record for that key. */
-                    break;
-                else
-                    /*
-                     * LAS key is greater than birthmark key, check next set of birthmark records.
-                     */
-                    continue;
-            }
+            /*
+             * Verify that our search_near has given us the right key.
+             *
+             * If it hasn't, this means that there are no longer any lookaside entries for that key.
+             * This is possible if the sweep server cleans some things up while the page was out of
```

Comment:
sweep can't go past this record since this has not been checkpointed yet. We should assert/panic if we cant find this key

Reply:
535a348